Adding case studies on learn mcp (#160)
microsoft/mcp-for-beginners: 09-CaseStudy/MCPforDocumentation.md

Modern development is more than just writing code—it’s about finding the right information at the right time. Documentation is everywhere, but it’s rarely where you need it most: inside your tools and workflows. By integrating documentation retrieval directly into your applications, you can save time, reduce context switching, and boost productivity. In this section, we’ll show you how to connect a Python client to the Microsoft Learn Docs MCP server, so you can access real-time, context-aware documentation without ever leaving your app.

We’ll walk through the process of establishing a connection, sending a request, and handling streaming responses efficiently. This approach not only streamlines your workflow but also opens the door to building smarter, more helpful developer tools.

## Learning Objectives

Why are we doing this? Because the best developer experiences are those that remove friction. Imagine a world where your code editor, chatbot, or web app can answer your documentation questions instantly, using the latest content from Microsoft Learn. By the end of this chapter, you’ll know how to:

- Understand the basics of MCP server-client communication for documentation

> - Implement a Python console application to connect to the Microsoft Learn Docs MCP server

Reviewer: ```suggestion
- Implement a Console application to connect to the Microsoft Learn Docs MCP server
```

- Use streaming HTTP clients for real-time documentation retrieval
- Log and interpret documentation responses in your application

You’ll see how these skills can help you build tools that are not just reactive, but truly interactive and context-aware.

## Assignment

Let’s put this into practice. Your task is to write a Python script that connects to the Microsoft Learn Docs MCP server, invokes the `microsoft_docs_search` tool, and logs the streaming response to the console. Why this approach? Because it’s the foundation for building more advanced integrations—whether you want to power a chatbot, an IDE extension, or a web dashboard.

You’ll find the code and instructions for this scenario in the `scenario3` folder within this case study. The steps will guide you through setting up the connection, making a query, and handling the results in a user-friendly way.

> - Write a Python client (`client.py`) that connects to the Microsoft Learn Docs MCP server at `https://learn.microsoft.com/api/mcp`

Reviewer: ```suggestion
- Write a client that connects to the Microsoft Learn Docs MCP server at `https://learn.microsoft.com/api/mcp`
```

- Use the official MCP Python SDK and streamable HTTP client for connection
- Call the `microsoft_docs_search` tool with a query parameter to retrieve documentation
- Implement proper logging and error handling 
- Create an interactive console interface to allow users to enter multiple search queries

## Sample Queries

Documentation is only useful if you can get to it quickly. Here are some example queries you can try in your interactive client to see just how powerful this integration can be:

- Tell me about Azure OpenAI from Microsoft Learn
- How do I deploy a web app using Azure App Service from Microsoft Learn?
- Show me the documentation for Azure Functions bindings from Microsoft Learn
- What is Azure Key Vault? Get details from Microsoft Learn
- Search for 'virtual network peering' in Microsoft Learn documentation

You can enter your own questions or documentation requests. Type `exit` to quit the client.

> 

Reviewer: Can you show an example of what it can look like to run the solution, e.g:
Prompt> my prompt
Answer> here's the results on topic NN

## Solution

> Now let’s see it in action. Below is a minimal sample solution to the assignment. The full code and details are available in the `scenario3/client.py` file. This example demonstrates the essential steps to connect to the Microsoft Learn Docs MCP server, call a tool, and print the result.

Reviewer: ```suggestion
Now let’s see it in action. Below is a minimal sample solution to the assignment. The full code and details are available here 
- [Python solution](./link to python)

This example demonstrates the essential steps to connect to the Microsoft Learn Docs MCP server, call a tool, and print the result.
```
